pkg/sql/set_cluster_setting.go

```diff
@@ -143,14 +204,27 @@ func (p *planner) SetClusterSetting(
 		return nil, err
 	}
 
-	if !p.execCfg.Codec.ForSystemTenant() {
+	if !forSystemTenant {
 		switch setting.Class() {
 		case settings.SystemOnly:
 			// The Lookup call above should never return SystemOnly settings if this
 			// is a tenant.
 			return nil, errors.AssertionFailedf("looked up system-only setting")
+
 		case settings.SystemVisible:
-			return nil, pgerror.Newf(pgcode.InsufficientPrivilege, "setting %s is only settable by the operator", name)
+			return nil, p.maybeAddSystemInterfaceHint(
+				pgerror.Newf(pgcode.InsufficientPrivilege, "setting %s is only settable by the operator", name),
+				"modify the cluster setting")
+		}
+	} else {
+		switch setting.Class() {
+		case settings.ApplicationLevel:
+			if !p.EvalContext().SessionData().Internal && RestrictAccessToSystemInterface.Get(&p.ExecCfg().Settings.SV) {
+				return nil, errors.WithHintf(
+					pgerror.Newf(pgcode.InsufficientPrivilege, "blocked update to application-level setting %s from the system interface", setting.Name()),
+					"Access blocked via %s to prevent likely user errors.\n"+
+						"Try changing the setting from a virtual cluster instead.", RestrictAccessToSystemInterface.Name())
+			}
```

Comment:
I worry a little that this is going to be too broad.  Some of these settings perhaps well fix to be SystemVisible, but settings like `enterprise.license`, `cluster.organization` need to be set in both clusters currently.